### PR TITLE
or-patterns: disallow in `let` bindings

### PR DESCRIPTION
**Pull request by mark-i-m**
~~Blocked on https://github.com/rust-lang/rust/pull/81869~~

Disallows top-level or-patterns before type ascription. We want to reserve this syntactic space for possible future generalized type ascription.

r? @petrochenkov

### CODE REVIEW

**Comment by mark-i-m** [comment]
Actually, I'm going to rebase this to not be based on the other PR... no need...

File: compiler/rustc_parse/src/parser/stmt.rs [hunk 1]
```diff
@@ -12,7 +12,10 @@ use rustc_ast::ptr::P;
 use rustc_ast::token::{self, TokenKind};
 use rustc_ast::util::classify;
 use rustc_ast::{AttrStyle, AttrVec, Attribute, MacCall, MacCallStmt, MacStmtStyle};
-use rustc_ast::{Block, BlockCheckMode, Expr, ExprKind, Local, Stmt, StmtKind, DUMMY_NODE_ID};
+use rustc_ast::{
+    Block, BlockCheckMode, Expr, ExprKind, Local, PatKind, Stmt, StmtKind, DUMMY_NODE_ID,
+};
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/stmt.rs:17]: Nit: could you avoid multi-line imports.

**Comment by petrochenkov** [comment]
I think this is indeed better blocked on https://github.com/rust-lang/rust/pull/81869 because the logic here should be shared with `fn parse_fn_param_pat`.

**Comment by petrochenkov** [comment]
`parse_fn_param_pat` is not used for closures, so or-patterns in both function parameters and let bindings can be parsed with `fn parse_pat_allow_top_alt` and then reported as errors post-factum.

**Comment by mark-i-m** [comment]
> parse_fn_param_pat is not used for closures, so or-patterns in both function parameters and let bindings can be parsed with fn parse_pat_allow_top_alt and then reported as errors post-factum.

Hmm.. so this made sense initially, but as I'm thinking about it, it seems that they are different... Specifically, in the case of `let`, I think we want to allow top-level or-patterns without a type, right? That requires first parsing the pattern and then looking for a `Colon` token after it... I think I can share the error generation itself, but not much more...

**Comment by petrochenkov** [comment]
I mean, in both cases we parse `parse_pat_allow_top_alt` and then look for a colon, except that it's mandatory in one case and optional in another.

**Comment by mark-i-m** [comment]
Hmm... I see. I'll take a closer look when I implement the changes in https://github.com/rust-lang/rust/pull/79278#issuecomment-778897785

**Comment by mark-i-m** [comment]
@petrochenkov I've pushed a commit which does a few things:
- Attempts to address your review comments.
- Changes `let` bindings to require `no_top_alt`, as per the Lang Team's request (https://github.com/rust-lang/rust/pull/79278#issuecomment-778897785)
- Creates a new feature gate `or_patterns_with_ty` that allows the use of top-level or-patterns in `let` if there is not type annotation.

Please let me know what you think.

**Comment by mark-i-m** [comment]
(I still need to bless tests and add a feature gate test, but I would appreciate if you could take a quick look at the overall direction.)

File: compiler/rustc_feature/src/active.rs [hunk 2]
```diff
@@ -638,6 +638,10 @@ declare_features! (
     /// Allows macro attributes to observe output of `#[derive]`.
     (active, macro_attributes_in_derive_output, "1.51.0", Some(81119), None),
 
+    /// Allows the use of top-level or-patterns in `let` if there is no type annotation.
+    /// Ex: `let (A | B) : Ty = foo;` or `let A | B = foo;` but not `let A | B : Ty;`.
+    (active, or_patterns_with_ty, "1.52.0", Some(54883), None),
```
Comment (petrochenkov) [inline compiler/rustc_feature/src/active.rs:643]: This is such an arbitrary and special-purpose feature gate.
I think it's better to just make this a hard error and suggest wrapping into parentheses.

`let P1 | P2 = foo` is a rare construction, so adding parentheses would require less symbols than writing the feature even on nightly.

File: compiler/rustc_parse/src/parser/item.rs [hunk 3]
```diff
@@ -1778,8 +1778,9 @@ impl<'a> Parser<'a> {
             let (pat, ty) = if is_name_required || this.is_named_param() {
                 debug!("parse_param_general parse_pat (is_name_required:{})", is_name_required);
 
-                let pat = this.parse_fn_param_pat()?;
-                if let Err(mut err) = this.expect(&token::Colon) {
+                let (pat, colon) = this.parse_fn_param_pat_colon()?;
+                if !colon {
+                    let mut err = this.expected_one_of_not_found(&[], &[token::Colon]).unwrap_err();
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/item.rs:1783]: Colon should already be in expected tokens, so simple `this.unexpected()` should work too.

**Comment by petrochenkov** [comment]
I'm skeptical about the decision to not stabilize `let P1 | P2 = E` (without type annotation), but it's a conservative error so it's not a big deal.

What I would worry about more is *nested* `|` patterns neighboring type ascription in situations like
```rust
let b @ P1 | P2: T = E;
```
I'm not sure whether code like this is impossible with the currently implemented parsing priorities, this needs an investigation.

**Comment by mark-i-m** [comment]
@petrochenkov Hi, sorry for the delay. I should have a lot more time over the next 2 weeks.

I think by construction of the grammar, it should not be possible to run into the case you mentioned. I went through the reference [on patterns](https://doc.rust-lang.org/nightly/reference/patterns.html), and as far as I can tell the only patterns that allow nested patterns and are undelimited are `@`, `&`, `&mut`, and `|`. Of these, `|` has lowest precedence (according to the RFC, and AFAICT the implementation too), so you would always get another error as far as I can tell: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3ed800cba5f8801d34d64195c5ebc681).

>  I think it's better to just make this a hard error and suggest wrapping into parentheses.

I'll remove the feature gate and rebase shortly. I'm remote though, so I'm not sure I'll be able to run the tests....

**Comment by petrochenkov** [comment]
@mark-i-m 
>I think by construction of the grammar, it should not be possible to run into the case you mentioned. I went through the reference [on patterns](https://doc.rust-lang.org/nightly/reference/patterns.html), and as far as I can tell the only patterns that allow nested patterns and are undelimited are `@`, `&`, `&mut`, and `|`. Of these, `|` has lowest precedence (according to the RFC, and AFAICT the implementation too), so you would always get another error as far as I can tell: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3ed800cba5f8801d34d64195c5ebc681).

Could you add tests for all these cases?

File: compiler/rustc_span/src/symbol.rs [hunk 4]
```diff
@@ -804,6 +804,7 @@ symbols! {
         options,
         or,
         or_patterns,
+        or_patterns_with_ty,
```
Comment (petrochenkov) [inline compiler/rustc_span/src/symbol.rs:807]: This is no longer necessary.

**Comment by mark-i-m** [comment]
I have blessed the tests (there is still one failure I don't understand with rustfix). I added the test src/test/ui/or-patterns/nested-undelimited-precedence.rs for the cases discussed above.

I also did some minor refactoring to consolidate code a bit more, and I had to update the unused parens lint.

**Comment by mark-i-m** [comment]
@petrochenkov I figured out why the remove-leading-vert test is failing on rustfix. I think it is because we are emitting multiple machine-applicable suggestions to remove the same `|`. The first one applies but the second one doesn't since we already removed the `|`. I'm not sure how to approach fixing that; it feels like something rustfix should handle better... Any ideas?

File: compiler/rustc_parse/src/parser/pat.rs [hunk 5]
```diff
@@ -128,36 +170,7 @@ impl<'a> Parser<'a> {
...
-                let msg = "wrap the pattern in parentheses";
-                let fix = format!("({})", pprust::pat_to_string(pat));
-                (msg, fix)
-            }
-        };
-
-        self.struct_span_err(pat.span, "an or-pattern parameter must be wrapped in parentheses")
-            .span_suggestion(pat.span, msg, fix, Applicability::MachineApplicable)
-            .emit();
+        self.parse_pat_before_ty(PARAM_EXPECTED, GateOr::No, RecoverComma::No, "parameters")
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/pat.rs:173]:
```suggestion
        self.parse_pat_before_ty(PARAM_EXPECTED, GateOr::No, RecoverComma::No, "function parameters")
```
Simply "parameters" sounds incomplete.

File: compiler/rustc_parse/src/parser/pat.rs [hunk 6]
```diff
@@ -104,12 +104,54 @@ impl<'a> Parser<'a> {
...
+        // We use `parse_pat_allow_top_alt` regardless of whether we actually want top-level
+        // or-patterns so that we can detect when a user tries to use it. This allows us to print a
+        // better error message.
+        let pat = self.parse_pat_allow_top_alt(expected, gate_or, rc)?;
+        let colon = self.eat(&token::Colon);
+
+        if let PatKind::Or(pats) = &pat.kind {
+            let msg = format!("top-level or-patterns are not allowed in {}", syntax_loc);
+            let (help, fix) = if pats.len() == 1 {
+                // If all we have a leading vert, then print a special message. This is the case if
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/pat.rs:131]:
```suggestion
                // If all we have is a leading vert, then print a special message. This is the case if
```

**Comment by petrochenkov** [comment]
@mark-i-m 
>I figured out why the remove-leading-vert test is failing on rustfix. I think it is because we are emitting multiple machine-applicable suggestions to remove the same `|`. The first one applies but the second one doesn't since we already removed the `|`. I'm not sure how to approach fixing that; it feels like something rustfix should handle better... Any ideas?

Not to suggest removing the same `|` multiple times?
I don't know anything about rustfix, honestly.

**Comment by mark-i-m** [comment]
@petrochenkov I think I have fixed it. It feels a bit messy, but I can't think of a better way to do it. Let me know what you think. I think this should be ready for review now (let me know if squashing before review would be helpful).

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

**Comment by estebank** [comment]
The code changes seem reasonable enough, but rustfix should be responsible for deduplicating repeated suggestions. What I have done in the past is not remove the suggestion, but remove _the entire error_ with `delay_as_bug()`. That has the benefit of not only helping rustfix but also of making the output less verbose, but the logic can get hairy _quickly_. But you already _have_ the logic here and arguably the following is redundant:

```

error: a trailing `|` is not allowed in an or-pattern
  --> $DIR/remove-leading-vert.rs:48:11
   |
LL |     let a | = 0;
   |         - ^ help: remove the `|`
   |         |
   |         while parsing this or-pattern starting here

error: top-level or-patterns are not allowed in `let` bindings
  --> $DIR/remove-leading-vert.rs:48:9
   |
LL |     let a | = 0;
   |         ^^^
```

**Comment by mark-i-m** [comment]
@estebank Thanks! I like that idea. Will push shortly.

I think the reason rustfix gets confused is that they are two equivalent but different suggestions:
- `a |` -> `a`
- `|` -> ` `

**Comment by petrochenkov** [comment]
Is https://github.com/rust-lang/rust/pull/82048/commits/95f4da783a1a294bfe36d7971268434a57cdf019 still necessary with `delay_span_bug`? It adds some noise that makes the logic harder to follow.

If it's necessary, could you some kind of `parse_pat_allow_top_alt_ext` for returning the pair and used only when necessary, so that all uses of `parse_pat_allow_top_alt` remain unaffected.

**Comment by mark-i-m** [comment]
@petrochenkov Unfortunately, it is necessary, as we only delay_span_bug if we haven't previously emitted a trailing-vert error. I've made the change you suggested.

**Comment by petrochenkov** [comment]
r=me after squashing commits.

**Comment by mark-i-m** [comment]
@petrochenkov Squashed!

**Comment by mark-i-m** [comment]
CI passing :)

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

**Comment by JohnTitor** [comment]
Failed in rollup: https://github.com/rust-lang/rust/pull/82840#issuecomment-792051784
Looks like we need a tweak to clippy.
@bors r-

**Comment by mark-i-m** [comment]
I think clippy should compile now.

**Comment by estebank** [comment]
@mark-i-m would you mind filing a ticket to change the unnecessary parens lint ahead of stabilization? I was recently [touching this area](https://github.com/rust-lang/rust/pull/82854/files) and should be straightforward to make it _not_ trigger for `let (A | B) = C;`.

**Comment by mark-i-m** [comment]
@estebank I had thought I already fixed the lint in this PR, though I'm not sure why it was firing for clippy...

**Comment by estebank** [comment]
@mark-i-m ah! Interesting.

**Pull request merged**